Always hash exists url

shtrom opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Documentation | no
| Translation   | no
| CHANGELOG.md  |no
Fixed tickets |#3919
| License       | MIT

Simplify the logic from #3158 by hashing all the urls from the request,
and only doing a search by hash. This allows to get performance benefits
from the new indexed hash column even when using older clients that do
not hash the URL in the request.

shtrom commented:
@j0k3r this is a small change in your hash_url branch. I still need to test it locally on my instance, but feedback is welcome.

shtrom commented:
Hum, Travis fails for my branch in the same pattern as it does for 2.4 without my change: https://travis-ci.org/wallabag/wallabag/jobs/524429166

shtrom commented:
Hum couple more thoughts: maybe we should hide the hash calculation and put it in `findByUrlAndUserId`, so any search actually uses the hash.

I do note, however, that the migration from #3158, while it creates the column, doesn't seem to calculate and populate the hashes for the existing entries. Could this be an issue, or is it handled somewhere else?

shtrom commented:
Actually, here's another change, making `EntryRepository::findByUrlAndUserId` hash the `url` under the hood, and call `EntryRepository::findByHasedUrlAndUserId` unconditionally.

This probably need some cleanup, as the `hash('sha1', ...)` is starting to be duplicated a bit too often, but what do you think of the idea?

One thing I am not sure about is what to do about the `urldecode` that `findByUrlAndUserId` was doing, but doesn't seem to be used in the `findByHashedUrlAndUserId` path. Or should the URL be `urldecode`d before being hashed?

shtrom commented:
Ok, tested on my instance, and it works nicely. No more slow queries, and Wallabagger is faster at turning green on already visited pages. This is mostly thanks to my second commit there, which uses the hash for everything. The question remains about the `urldecode` part. It doesn't seem to be used consistently.

Another thing I did notice is that the `wallabag:generate-hashed-urls` command is not ran during the migration, so it would probably be worth adding a note that this needs to happen in the release notes.

j0k3r commented on src/Wallabag/CoreBundle/Resources/config/services.yml:
```diff
@@ -220,6 +220,13 @@ services:
             - "%wallabag_core.site_credentials.encryption_key_path%"
             - "@logger"
 
+    wallabag_core.helper.url_hasher:
```
I'm wondering if this should be a service or a simple class with static method in it :thinking: 
BTW, the file is missing `UrlHasher.php`

shtrom replied:
:man_facepalming:

shtrom replied:
I thought about a static class, then figured that we may want to be able to change the hash algorithm later on at runtime, so made that a configurable instance attribute.

j0k3r replied:
For now I do not think we'll need to update the algorithm at runtime and I prefer to keep that helper dead simple with a static method and no service.

`sha1` can still be the default algo.

```php
<?php

namespace Wallabag\CoreBundle\Helper;

/**
 * Hash URLs for privacy and performance.
 * Internal helper to centralize how the url is hashed.
 */
class UrlHasher
{
    /**
     * @param string $url
     * @param string $algorithm Available algo for the PHP function `hash`
     *
     * @return string hashed $url
     */
    public function hash(string $url, string $algorithm = 'sha1')
    {
        return hash($algorithm, $url);
    }
}
```

shtrom commented:
There we go. I added a `UrlHasher` helper, with a configurable hash algorithm (defaulting to `sha1` in the constructor). I roped it in in the `services.yml`, though I haven't found a way to make it fetchable by `Entry`s; those instantiate the class directly, and rely on the default hash algorithm to be the same as for the other services (I added a note to that effect in the `services.yml`).

shtrom commented:
How about now? A static class does limit the changes needed quite a bit :+1:

j0k3r commented:
I rebased, squashed some commits & fix tests.
I think this is ok now.

j0k3r commented:
Thanks for starting the job @shtrom !
